### ISSUE TITLE
Error in training: assert maskp[:, i].sum().item() != 0

### ISSUE THREAD

**Issue by ylxbyy** [zhou13/neurvps#11]
Hello~ 

Thanks a lot for sharing this work.

I met the following error when training the network. The problem seems to be that there are no positive results in the sampled vanishing points in a M.multires’s iteration, so “maskp[:, i].sum().item() != 0” cannot be satisfied. But the network already ran 16 epochs without such errors. Why did it happen here (I tried to train again and met the error almost at the same epoch) and how to avoid such error in training?

Could you please give me some help in solving the problem? 

Some records in the console window:
......
014/0600k| 26.52771| 0.10037| 6.36750| 0.08514| 7.50082| 0.08439| 7.30224| 0.07944| 4.27890| 0.09493| 0.63400        
015/0624k| 24.97396| 0.10929| 6.02662| 0.08992| 7.15550| 0.08672| 6.84507| 0.08786| 3.89247| 0.10852| 0.57199        
015/0648k| 25.57682| 0.09436| 6.31269| 0.07620| 7.38553| 0.07375| 7.02303| 0.08718| 3.88921| 0.11177| 0.52310        
016/0672k| 23.83661| 0.10088| 6.06711| 0.07959| 6.97335| 0.07531| 6.43446| 0.09949| 3.44505| 0.09952| 0.46185        
016/0696k| 19.81437| 0.12112| 4.67192| 0.09953| 5.51086| 0.09114| 5.40486| 0.07137| 3.29440| 0.08435| 0.46482        
016/0698k| 1.29788| 0.35235| 0.20292| 0.26271| 0.11353| 0.13166| 0.07957| 0.05743| 0.03989| 0.02206| 0.03577| 04.1 Traceback (most recent call last):
  File "/home/yyz/neurvps/train_gsv.py", line 194, in <module>
    main()
  File "/home/yyz/neurvps/train_gsv.py", line 186, in main
    trainer.train()
  File "/home/yyz/neurvps/neurvps/trainer.py", line 253, in train
    self.train_epoch()
  File "/home/yyz/neurvps/neurvps/trainer.py", line 171, in train_epoch
    result = self.model(input_dict)
  File "/home/yyz/anaconda3/envs/neurvps/lib/python3.7/site-packages/torch/nn/modules/module.py", line 532, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/yyz/anaconda3/envs/neurvps/lib/python3.7/site-packages/torch/nn/parallel/data_parallel.py", line 150, in forward
    return self.module(*inputs[0], **kwargs[0])
  File "/home/yyz/anaconda3/envs/neurvps/lib/python3.7/site-packages/torch/nn/modules/module.py", line 532, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/yyz/neurvps/neurvps/models/vanishing_net.py", line 71, in forward
    assert maskp[:, i].sum().item() != 0
AssertionError

**Comment by zhou13** [comment]
Since you are training with a modified neurvps with `train_gsv.py` (maybe on your own dataset), I cannot provide much support. I will suggest you pdb into the assertation error and see what causes the problem. Normally the vp sampling code should generate both positive and negative vp, so you need to check why this is not the case anymore.

**Comment by ylxbyy** [comment]
Thank you for you reply and suggestion.

**Comment by ylxbyy** [comment]
@zhou13 

Hello, Mr. Zhou, 

I have debugged into the assertion error and find that y[:, 0] == 0 so maskp[:, 0].sum().item() == 0. But it is strange that: (1) all 33 samples are not positive in the [0, 0.0013457768043554] threshold; (2) the assertion error usually occur at the 16th epoch (1488308 iterations this time; I have run the training process several times, and the condition almost the same). 

In the debug mode, when I did the following test when it ran to the assertion:

>>>aa = sample_sphere(vpts_gt[0,1,:], 0, 0.0013457768043554)
>>>to_label(aa, vpts_gt[0,1,:])
[1, 1, 1, 1, 1]
>>>aa = sample_sphere(vpts_gt[0,2,:], 0, 0.0013457768043554)
>>>to_label(aa, vpts_gt[0,2,:])
[0, 1, 1, 1, 1]
>>>aa
array([-0.83703564,  0.19581235,  0.51090984])
>>>vpts_gt[0,2,:]
array([-0.83775014,  0.19568816,  0.50978506], dtype=float32)
>>>np.min(np.arccos(np.abs(vpts_gt[0,2,:] @ aa).clip(max=1)))
0.001356664619915869
>>>bb = sample_sphere(vpts_gt[0,2,:], 0, 0.0013457768043554)
>>>to_label(bb, vpts_gt[0,2,:])
[1, 1, 1, 1, 1]
>>>to_label(sample_sphere(vpts_gt[0,2,:], 0, 0.0013457768043554), vpts_gt[0,2,:])
[1, 1, 1, 1, 1]
>>>to_label(sample_sphere(vpts_gt[0,2,:], 0, 0.0013457768043554), vpts_gt[0,2,:])
[1, 1, 1, 1, 1]
>>>to_label(sample_sphere(vpts_gt[0,2,:], 0, 0.0013457768043554), vpts_gt[0,2,:])
[1, 1, 1, 1, 1]
>>>to_label(sample_sphere(vpts_gt[0,2,:], 0, 0.0013457768043554), vpts_gt[0,2,:])
[1, 1, 1, 1, 1]

The condition actually happened that a sample created within a threshold (regarded as positive) using the ‘sample_sphere’ function has an error larger than the threshold that was computed by the ‘to_label’ function (0.001356664619915869>0.0013457768043554). 

So I wonder:
(1)	Whether it is because of truncation error and will it be useful to change the data type to float64 (now is float32)?
(2)	As such condition often happens after about two days’ training (16 epoch), is it because some kind of errors accumulated in the computing? 
(3)	Besides, changing the ‘1e-6’ value in the ‘to_label’ function (return [int(degree < res + 1e-6) for res in M.multires]) to a larger value seems to be a solution of the assertion error. But if the value becomes larger, it may influence the judgement of negative samples. What value will you suggest for balancing the above two?

Thank you.
